pkg/controllermanager/controller/backupbucket/backup_bucket_reconciler.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	controllerutils "github.com/gardener/gardener/pkg/controllermanager/controller/utils"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/operation/common"
+	utilerrors "github.com/gardener/gardener/pkg/utils/errors"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// reconciler implements the reconcile.Reconcile interface for backupBucket reconciliation.
+type reconciler struct {
+	ctx      context.Context
+	client   client.Client
+	recorder record.EventRecorder
+	logger   *logrus.Logger
+}
+
+// newReconciler returns the new backupBucker reconciler.
+func newReconciler(ctx context.Context, gardenClient client.Client, recorder record.EventRecorder) reconcile.Reconciler {
+	return &reconciler{
+		ctx:      ctx,
+		client:   gardenClient,
+		recorder: recorder,
+		logger:   logger.Logger,
+	}
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	bb := &gardencorev1alpha1.BackupBucket{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, bb); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Debugf("[BACKUPBUCKET RECONCILE] %s - skipping because BackupBucket has been deleted", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Infof("[BACKUPBUCKET RECONCILE] %s - unable to retrieve object from store: %v", request.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	if bb.Spec.Seed == nil {
+		message := "Cannot reconcile BackupBucket: Waiting for BackupBucket to get scheduled on a Seed"
+		r.recorder.Event(bb, corev1.EventTypeWarning, "OperationPending", message)
+		return reconcile.Result{}, utilerrors.WithSuppressed(fmt.Errorf("backupBucket %s has not yet been scheduled on a Seed", bb.Name), r.updateBackupBucketStatusPending(bb, message))
+	}
+
+	if bb.DeletionTimestamp != nil {
+		return r.deleteBackupBucket(bb)
+	}
+	// When a BackupBucket deletion timestamp is not set we need to create/reconcile the backup bucket.
+	return r.reconcileBackupBucket(bb)
+}
+
+func (r *reconciler) reconcileBackupBucket(backupBucket *gardencorev1alpha1.BackupBucket) (reconcile.Result, error) {
+	backupBucketLogger := logger.NewFieldLogger(logger.Logger, "backupbucket", fmt.Sprintf("%s", backupBucket.Name))
+	if updateErr := r.updateBackupBucketStatusProcessing(backupBucket, "Reconciliation of Backup Bucket state in progress.", 2); updateErr != nil {
+		backupBucketLogger.Errorf("Could not update the BackupBucket status after reconciliation start: %+v", updateErr)
+		return reconcile.Result{}, updateErr
+	}
+
```

Comment:
As mentioned above, I think we should add a finalizer to the referenced secret here.
Also, we should add the finalizer to the `BackupBucket` resource here. It's actually not a good idea to let the apiserver add the finalizer. We should change this also for the other resources like Shoot, etc. But not with this PR. Let's now just do it "correctly" for the newly introduced resources and fix the other ones separately.

Reply:
Sure.